### PR TITLE
IH-676: Improve xe command line autocompletion

### PR DESCRIPTION
**Pull request by last-genius**
Best reviewed by commit. No functional changes - these are all in the bash autocompletion script, which gets triggered on `<TAB><TAB>` and at worst just slightly ruins the user experience.

Additions are described in commits, but these are the most prominent things to test (with alternative similar commands as well - I might have missed something):

To test this, do:
```
$ scp ocaml/xe-cli/bash-completion root@YOUR_HOST:/etc/bash_completion.d/xe
```

1. Grouping
```
$ xe <TAB>
appliance-     feature-       pgpu-          sm-            vbd-
audit-         gpu-           pif-           snapshot-      vdi-
bond-          host-          pool-          sr-            vgpu-
cd-            lvhd-          pusb-          subject-       vif-
certificate-   message-       pvs-           task-          vlan-
cluster-       network-       repository-    template-      vm-
console-       observer-      role-          tunnel-        vmss-
diagnostic-    patch-         sdn-           update-        vtpm-
drtask-        pbd-           secret-        usb-           vusb-
event-         pci-           session-       user-

$ xe vm<TAB>
vm-     vmss-

$ xe host-data-source-query data-source=<TAB>
cpu0                                 memory_total_kib
cpu0-                                pif_aggr_rx
CPU0-                                pif_aggr_tx
cpu1                                 pif_eth0_rx
cpu1-                                pif_eth0_rx_errors
CPU1-                                pif_eth0_tx
cpu2                                 pif_eth0_tx_errors
cpu2-                                pif_eth1_rx
CPU2-                                pif_eth1_rx_errors
cpu3                                 pif_eth1_tx
cpu3-                                pif_eth1_tx_errors
CPU3-                                pool_session_count
cpu4                                 pool_session_creation_rate
cpu4-                                pool_task_count
```

2. Integrated help strings
```
$ xe vmss-<TAB>
vmss-create        -  Create a VM snapshot schedule
vmss-destroy       -  Destroy a VM snapshot schedule
vmss-list          -  Lists all the vmsss, filtering on the optional arguments
vmss-param-add     -  Adds to a set or map parameter
vmss-param-clear   -  Clears the specified parameter (param-name can be name-label,name-description,enabled,type,retained-snapshots,frequency,schedule,VMs)
vmss-param-get     -  Gets the parameter specified of the object
vmss-param-list    -  Lists all the parameters of the object specified by the uuid
vmss-param-remove  -  Removes a member or a key,value pair from a set/map respectively
vmss-param-set     -  Sets the parameter specified
```

3. Autocompletion for `xe help`

4. Auxiliary information is displayed next to UUIDs
```
$ xe vdi-create sr-uuid=<TAB>
44f0e162-6c9e-2abf-2b15-d11822314283  - DVD drives
8186b4b6-8913-97c5-2975-fafecc63f48e  - Local storage
8fcdf752-aa0d-3caa-4ccd-1d9b9bdd50d5  - Removable storage
a145c6d8-0eed-adb1-476d-a35a2906d6e4  - XenRT Windows ISOs
cfb8d089-b0dd-764d-238c-fc0e96c77dcc  - XenRT Linux ISOs

$ xe network-param-add uuid=<TAB>
1e5e2e82-c562-e3f8-b635-02c7c6ce7be3  - Pool-wide network associated with eth0
3161c966-f307-c692-378e-36901026771b  - Host internal management network
ad57184f-cd27-7401-fbda-4707822127e9  - Pool-wide network associated with eth1

$ xe bond-create pif-uuids=<TAB>
7819fbb5-537b-85ea-e31c-96424fdc9349  - eth0
78dfe620-e083-e0c0-5426-e0664e1b5443  - eth1
```
5. Colon in records is treated ok:
```
$ xe vmss-create <TAB>
enabled=             name-description=    retained-snapshots=  type=
frequency=           name-label=          schedule:

$ xe vmss-create schedule:<TAB>
days=  hour=  min=

$ xe vmss-create schedule:days=<TAB>
friday      monday      saturday    sunday      thursday    tuesday     wednesday
```

6. Autocompletion for boolean parameters
```
$ xe pif-list physical=<TAB>
false   true
```

7. No suggestions for illegal parameters

Before:
```
$ xe pif-list sr-uuid=<TAB>
44f0e162-6c9e-2abf-2b15-d11822314283   a145c6d8-0eed-adb1-476d-a35a2906d6e4
8186b4b6-8913-97c5-2975-fafecc63f48e   cfb8d089-b0dd-764d-238c-fc0e96c77dcc
8fcdf752-aa0d-3caa-4ccd-1d9b9bdd50d5
```
After:
```
$ xe pif-list sr-uuid=<TAB>
<EMPTY>
```

8. Previously entered parameters will not be suggested further:
```
$ xe vdi-create <TAB>
name-label=    sm-config:     tags:          virtual-size=
sharable=      sr-uuid=       type=

$ xe vdi-create sr-uuid=<TAB>
44f0e162-6c9e-2abf-2b15-d11822314283  - DVD drives
8186b4b6-8913-97c5-2975-fafecc63f48e  - Local storage
8fcdf752-aa0d-3caa-4ccd-1d9b9bdd50d5  - Removable storage
a145c6d8-0eed-adb1-476d-a35a2906d6e4  - XenRT Windows ISOs
cfb8d089-b0dd-764d-238c-fc0e96c77dcc  - XenRT Linux ISOs

$ xe vdi-create sr-uuid=44f0e162-6c9e-2abf-2b15-d11822314283 <TAB>
name-label=    sm-config:     type=
sharable=      tags:          virtual-size=
```

9. Required and optional parameters are distinguished explicitly:
```
$ xe vmss-create <TAB>
OPTIONAL: enabled=             REQUIRED: frequency=
OPTIONAL: name-description=    REQUIRED: name-label=
OPTIONAL: retained-snapshots=  REQUIRED: type=
OPTIONAL: schedule:
$ xe vmss-create n<TAB>
OPTIONAL: name-description=  REQUIRED: name-label=
```

10. A power user "cheat-code" shortcut that inserts all required parameter names:
```bash {1|all}
$ xe vdi-create <Ctrl-r><q>

$ xe vdi-create sr-uuid= name-label= virtual-size=
```

-----

**Things that are not possible within the constraints of Bash autocompletion:**
* Better comma handling, as in `xe pif-list params=uuid,VLAN,`, taking zsh's autocompletion for filenames in `{}` as an example
* Colors in autocompletion

**Things that will be improved with a new version of Bash:**
* Required and optional parameters can be represented in a better way once `complete` gets an `-o nosort` option (Bash 4.4):
```
$ xe vmss-create <TAB><TAB>
* REQUIRED PARAMETERS:
name-label= type= frequency=
* OPTIONAL PARAMETERS:
enabled= retained-snapshots= schedule: name-description=
```

**Things that depend on rearchitecting the bash autocompletion into a shell-like solution or an SDK-like solution:**
* Complete and up-to-date knowledge of types, records and the nuances in how these are parsed.

### CODE REVIEW

**Comment by last-genius** [comment]
Found a little "bug", so turning this into a draft in the meantime

File: ocaml/xe-cli/bash-completion [hunk 1]
```diff
@@ -140,8 +142,6 @@ _xe()
                         cmd=vm-list;;
                     diagnostic-vdi-status)
                         cmd=vdi-list;;
-                    host-cpu-info)
-                        cmd=host-list;;
```
Comment (psafont) [inline ocaml/xe-cli/bash-completion:104]: Was this removed on purpose? It's not mentioned in the commit message

File: ocaml/xe-cli/bash-completion [hunk 2]
```diff
@@ -280,15 +277,12 @@ _xe()
...
                     set_completions 'Disk,CD' "$value"
-                    return 0
                 elif [[ "${fst}" == "vdi" ]]; then
                     set_completions 'system,user,suspend,crashdump' "$value"
-                    return 0
                 elif [[ "${fst}" == "sr" ]]; then
-                    local words=$( ${xe} sm-list params=type --minimal 2>/dev/null )
-                    set_completions "$words" "$value"
-                    return 0
+                    set_completions_for_names "sm-list" "type" "$value"
```
Comment (psafont) [inline ocaml/xe-cli/bash-completion]: This change doesn't seem related to the cleanup either

**Review by psafont — COMMENTED** [review]
Found a couple of changes that were introduced in the first commit, but should be part of a later one.

The code is complex, but it's easily verifiable. Knowing that it only affects the autocompletion should also makes it easier to merge

File: ocaml/xe-cli/bash-completion (hunk 1 above)
Comment (last-genius) [reply ocaml/xe-cli/bash-completion:104]: this case is covered by the wildcard host-cpu-* just below

Comment (last-genius) [reply ocaml/xe-cli/bash-completion:104]: Oops, not really. I'll revert this particular change.

File: ocaml/xe-cli/bash-completion (hunk 2 above)
Comment (last-genius) [reply ocaml/xe-cli/bash-completion]: it's just cleaning up duplicate code. the function does these commands already.

**Comment by lindig** [comment]
The correctness is difficult to review but it is not crucial because you can always provide full arguments and not rely on completion. We need to be sure that this is still efficient in large pools and does not run any malicious code as part of computing the completion. Otherwise I suggest to merge this and to try it.

**Review by psafont — APPROVED** [review]
As said before, the changes are complex, but the scope for adding issues is quite limited

**Review by lindig — APPROVED** [review]
